sema4ai/src/sema4ai_code/action_server.py

```diff
@@ -229,3 +236,282 @@ def get_error_message(result_message: str) -> str:
+                    message=f"Error parsing output.\n{command_result.result}",
+                )
+            return ActionResult(success=True, message=None, result=result)
+        else:
+            return ActionResult(
+                success=False,
+                message=f"Error building action package.\n{command_result.message or ''}",
+            )
+
+    def package_upload(
+        self,
+        package_path: str,
+        organization_id: str,
+        access_credentials: Optional[str],
+        hostname: Optional[str],
+    ) -> ActionResult[ActionServerPackageUploadStatusDict]:
+        """
+        Upload action package to Control Room.
+
+        Args:
+            package_path, path tot he package to upload.
+            organization_id, Control Room organization id.
+        """
+        args = [
+            "package",
+            "upload",
+            "--package-path",
+            package_path,
+            "--organization-id",
+            organization_id,
+            "--json",
+        ]
+
+        if access_credentials and hostname:
+            args.extend(
+                ["--access-credentials", access_credentials, "--hostname", hostname]
+            )
+
+        command_result = self._run_action_server_command(
+            args, timeout=ONE_MINUTE_S * 10
+        )
+
+        if command_result.success:
+            try:
+                if not command_result.result:
+                    return ActionResult(
+                        success=False,
+                        message="Error not output from action server.",
+                    )
+                result = json.loads(command_result.result)
+            except Exception:
+                return ActionResult(
+                    success=False,
+                    message=f"Error parsing output.\n{command_result.result}",
+                )
+            return ActionResult(success=True, message=None, result=result)
+        else:
+            return ActionResult(
+                success=False,
+                message=f"Error uploading package to Control Room.\n{command_result.message or ''}",
+            )
+
+    def package_status(
+        self,
+        package_id: str,
+        organization_id: str,
+        access_credentials: Optional[str],
+        hostname: Optional[str],
+    ) -> ActionResult[ActionServerPackageUploadStatusDict]:
+        """
+        Get uploaded action package status from Control Room.
+
+        Args:
+            package_id, uploaded package id.
+            organization_id, Control Room organization id.
+        """
+        args = [
+            "package",
+            "status",
+            "--package-id",
+            package_id,
+            "--organization-id",
+            organization_id,
+            "--json",
+        ]
+
+        if access_credentials and hostname:
+            args.extend(
+                ["--access-credentials", access_credentials, "--hostname", hostname]
+            )
```

Comment:
Same question as changelog.

Reply:
They are optional, the caller controls if they are needed. Action server don't need them if user has already logged in.